lib/SPIRV/libSPIRV/SPIRVValue.h

```diff
@@ -175,40 +175,59 @@ template <spv::Op OC> class SPIRVConstantBase : public SPIRVValue {
   float getFloatValue() const { return Union.FloatVal; }
   double getDoubleValue() const { return Union.DoubleVal; }
   unsigned getNumWords() const { return NumWords; }
-  SPIRVWord *getSPIRVWords() { return Union.Words; }
+  const std::vector<SPIRVWord> &getBigValWords() { return BigValWords; }
 
 protected:
+  constexpr static SPIRVWord FixedWC = 3;
+  constexpr static unsigned WordBitWidth = sizeof(SPIRVWord) * CHAR_BIT;
+  constexpr static unsigned UnionWordWidth = 2;
+
   void recalculateWordCount() {
-    NumWords = (Type->getBitWidth() + 31) / 32;
-    WordCount = 3 + NumWords;
+    NumWords = (Type->getBitWidth() + WordBitWidth - 1) / WordBitWidth;
+    WordCount = FixedWC + NumWords;
   }
   void validate() const override {
     SPIRVValue::validate();
-    assert(NumWords >= 1 && NumWords <= 64 && "Invalid constant size");
+    assert(NumWords >= 1 && "Invalid constant size");
   }
   void encode(spv_ostream &O) const override {
     getEncoder(O) << Type << Id;
-    for (unsigned I = 0; I < NumWords; ++I)
-      getEncoder(O) << Union.Words[I];
+    if (NumWords <= UnionWordWidth) {
+      for (unsigned I = 0; I < NumWords; ++I)
+        getEncoder(O) << Union.Words[I];
+    } else {
+      for (const auto &Word : BigValWords)
+        getEncoder(O) << Word;
+    }
   }
   void setWordCount(SPIRVWord WordCount) override {
     SPIRVValue::setWordCount(WordCount);
-    NumWords = WordCount - 3;
+    NumWords = WordCount - FixedWC;
   }
   void decode(std::istream &I) override {
     getDecoder(I) >> Type >> Id;
-    for (unsigned J = 0; J < NumWords; ++J)
-      getDecoder(I) >> Union.Words[J];
+    if (NumWords <= UnionWordWidth) {
+      for (unsigned J = 0; J < NumWords; ++J)
+        getDecoder(I) >> Union.Words[J];
+    } else {
+      BigValWords.resize(NumWords);
+      for (auto &Word : BigValWords)
+        getDecoder(I) >> Word;
+    }
   }
 
   unsigned NumWords;
   union UnionType {
     uint64_t UInt64Val;
     float FloatVal;
     double DoubleVal;
-    SPIRVWord Words[64];
+    SPIRVWord Words[UnionWordWidth];
     UnionType() { UInt64Val = 0; }
   } Union;
+
+private:
+  // For storing values of any bitwidth greater than 64.
+  std::vector<SPIRVWord> BigValWords;
```

Comment:
Is it possible to use it for small integers too and get rid of the union?

Reply:
The union is needed to be able to quickly reinterpret constant as `float`, `int` or `double`, so it would require some refactoring in methods like `getFloatValue()` to get rid of it.

However, I do agree that we could just put the vector into the union to unify handling of huge integers with the handling of the rest of types

Reply:
Refactored code.